src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ActionPlanService.kt

```diff
@@ -0,0 +1,70 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.service
+
+import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.config.Code
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.config.FieldError
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.config.ValidationError
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.dto.DraftActionPlanDTO
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ActionPlan
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.ActionPlanActivity
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.AuthUser
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ActionPlanRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.AuthUserRepository
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.ReferralRepository
+import java.time.OffsetDateTime
+import java.util.UUID
+import java.util.UUID.randomUUID
+
+@Service
+class ActionPlanService(
+  val authUserRepository: AuthUserRepository,
+  val referralRepository: ReferralRepository,
+  val actionPlanRepository: ActionPlanRepository
+) {
+
+  fun createDraftActionPlan(
+    referralId: UUID,
+    numberOfSessions: Int?,
+    activities: List<ActionPlanActivity>,
+    createdByUser: AuthUser
+  ): ActionPlan {
+
+    val draftActionPlan = ActionPlan(
+      id = randomUUID(),
+      numberOfSessions = numberOfSessions,
+      createdBy = authUserRepository.save(createdByUser),
+      createdAt = OffsetDateTime.now(),
+      referral = referralRepository.getOne(referralId),
+      activities = activities
+    )
+
+    return actionPlanRepository.save(draftActionPlan)
+  }
+
+  fun getDraftActionPlan(id: UUID): ActionPlan? {
+    return actionPlanRepository.findByIdAndSubmittedAtIsNull(id)
+  }
+
+  fun updateActionPlan(actionPlan: ActionPlan, update: DraftActionPlanDTO): ActionPlan {
+    validateDraftActionPlanUpdate(update)
```

Comment:
Just a thought but if the number of sessions is the only attribute that can be updated then should the API not reflect that. Also, DTOs are a concern of the controllers and they convert them into entities for the service business tier to act on.

fun updateActionPlan(id: UUID, numberOfSessions: Int): ActionPlan 

Also, the existing version of the plan should be retrieved and not supplied as the controller may inadvertently change other fields it's not supposed to as part of the update. Also, isn't the missing draft action plan check part of the business logic and therefore be done as part of this business method.